src/watch.ts

```diff
@@ -153,6 +162,70 @@ export class Watch {
     }, config.retrySkipped);
   }
 
+  /**
+   * Regularly try to refresh packages informations.
+   * Mostly here for time based data: download stats, popularity, etc...&.
+   */
+  async checkToRefresh(): Promise<void> {
+    log.info('Checking refresh jobs');
+
+    // schedule next iteration
+    setTimeout(() => {
+      this.checkToRefresh();
+    }, config.refreshPeriod);
+
+    // We list all values in facet and pick the oldest (hopefully the oldest is in the list)
+    const res = await this.mainIndex.search('', {
+      facets: ['_searchInternal.expiresAt'],
+      hitsPerPage: 0,
+      sortFacetValuesBy: 'alpha',
+    });
+    const list = Object.keys(res.facets!['_searchInternal.expiresAt']).sort();
+    log.info(' > Found', list.length, 'expiration values');
+    if (list.length <= 0) {
+      return;
+    }
+
+    const pick = list.shift()!;
+    const expiresAt = new Date(parseInt(pick, 10));
+    if (expiresAt.getTime() > Date.now()) {
+      log.info(' > Oldest date is in the futur');
+      return;
+    }
+    log.info(' > Picked the oldest', expiresAt.toISOString());
+
+    // Retrieve some packages to update, not too much to avoid flooding the queue
+    const pkgs = await this.mainIndex.search<FinalPkg>('', {
+      facetFilters: [`_searchInternal.expiresAt:${pick}`],
+      facets: ['_searchInternal.expiresAt'],
+      hitsPerPage: 20,
+    });
+    log.info(' > Found', pkgs.hits.length, 'expired packages');
+
+    const pushed: string[] = [];
+    for (const pkg of pkgs.hits) {
+      if (!pkg.rev) {
+        continue;
+      }
+      pushed.push(pkg.objectID);
+
+      // If an update come at the same time, this could override the update with old rev.
```

Comment:
do you mean while this function is called, or right after this function is called? Could we use a semaphore, and delay adding items to changes consumer until this process is finished, to make sure it's first? Or maybe read the changes that happen between start of this process and end of this process, and skip adding if it's already refreshed for another reason?